### PR TITLE
BUG: declare `np.unstack` as subclass-safe (fix incompatibility with NumPy 2.1)

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Fixes an incoming incompatibility ([example logs](https://github.com/astropy/astropy/actions/runs/9730358857/job/26853352422))
xref: https://github.com/numpy/numpy/pull/26579

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
Thanks! LGTM but I'll let subpackage maintainers approve.

**Comment by neutrinoceros** [comment]
I'll take it your approval is enough for `utils` 😸 
Let's ping @mhvk and @nstarman for units

**Comment by pllim** [comment]
`utils.masked` is beyond me...

File: astropy/units/quantity_helper/function_helpers.py
Comment (nstarman) [inline astropy/units/quantity_helper/function_helpers.py:117]: Can't this remain an `else`?

File: astropy/utils/masked/function_helpers.py
Comment (nstarman) [inline astropy/utils/masked/function_helpers.py:144]: Same here?

File: astropy/units/quantity_helper/function_helpers.py
Comment (neutrinoceros) [reply astropy/units/quantity_helper/function_helpers.py:117]: It could, but it's bound to take this form sooner or later, when numpy<2 is dropped. I also find that if/else becomes more confusing as soon as there are more than 2 cases to deal with. If you're not convinced, please feel free to push to this branch, as I won't be able to until tomorrow morning.

Comment (neutrinoceros) [reply astropy/units/quantity_helper/function_helpers.py:117]: Wait I was wrong. It doesn't need to evolve as I described, we'll just deindent it when it's not conditional anymore. My second point still stands though.

Comment (nstarman) [reply astropy/units/quantity_helper/function_helpers.py:117]: The second `if` (for ` not NUMPY_LT_2_1`) shouldn't be changed. So this will remain a 2-case if/else. 

```python
if NUMPY_LT_2_0: ...
else: ...
if not NUMPY_LT_2_1: ...
```

Which in a few years will become

```python
...
if not NUMPY_LT_2_1: ...
```

Comment (neutrinoceros) [reply astropy/units/quantity_helper/function_helpers.py:117]: Triple `if` is still slightly more appealing to me but I don't hold strong opinions here.

Comment (mhvk) [reply astropy/units/quantity_helper/function_helpers.py:117]: I'm with @nstarman here - prefer the `else`, but so slightly that one might as well not run CI for it again.

**Review by mhvk — APPROVED** [review]
Thanks!

**Pull request merged**
